lib/pacemaker/pcmk_cluster_queries.c

```diff
@@ -334,16 +345,26 @@ pcmk__designated_controller(pcmk__output_t *out, guint message_timeout_ms)
         .message_timeout_ms = message_timeout_ms,
         .pcmkd_state = pcmk_pacemakerd_state_invalid,
     };
-    pcmk_ipc_api_t *controld_api = ipc_connect(&data, pcmk_ipc_controld, designated_controller_event_cb);
+    enum pcmk_ipc_dispatch dispatch_type = pcmk_ipc_dispatch_main;
+    pcmk_ipc_api_t *controld_api = NULL;
+
+    if (message_timeout_ms == 0) {
+        dispatch_type = pcmk_ipc_dispatch_sync;
```

Comment:
AFAICT this also changed the behavior of "crmadmin -D" under certain circumstance.

Now upon cluster startup, if "crmadmin -D" is executed before a DC is elected yet, it will be hanging forever .

AFAICS the lookup involves a ping request to the DC. But given that there's not a DC to actually handle the request at the moment,  crmadmin most probably won't receive any reply under the situation...

Previously with an implied timeout 30s, the command would eventually return with an error though rather than hanging forever.

Or course first it's a question what should be the best as the default for such a case.

Reply:
The idea was that synchronous requests are supposed to time out after 5 seconds...

> @nrwahl2 , do you want to look into that?

"want" is such a loaded word ;)

Reply:
https://projects.clusterlabs.org/T735

Reply:
> https://projects.clusterlabs.org/T735

Good thinking over there. 

My two cents. Unless I'm missing anything, it sounds good enough to me for "crmadmin -D" to have the same understanding as the local controller in regard of who is the DC. If our controller doesn't think there's a DC yet, probably so be it and just tell crmadmin the same? 

Basically it probably doesn't really need to cost a ping request to DC to know who is the DC. It would be more meaningful if we'd like to achieve the following with a single command/request:

`crmadmin -S $(crmadmin -D -q)`

Reply:
> Now upon cluster startup, if "crmadmin -D" is executed before a DC is elected yet, it will be hanging forever .
> 
> AFAICS the lookup involves a ping request to the DC. But given that there's not a DC to actually handle the request at the moment, crmadmin most probably won't receive any reply under the situation...

Another exposed issue is, without anything to read from the IPC at the moment, the looping in pcmk__send_ipc_request() becomes extremely busy since crm_ipc_read() keeps getting -EAGAIN:

https://github.com/ClusterLabs/pacemaker/blob/main/lib/common/ipc_client.c#L703-L706

Maybe a sleep()  or pcmk_poll_ipc() before retry would make more sense?

Reply:
`sleep()` would still be synchronous, and I think `pcmk_poll_ipc()` would be just as busy -- it's just shifting the work to the `poll()` system call. It blocks on `poll()` until either we receive something or the timeout we pass expires. I mentioned in T735 that at a glance, it looks like `pcmk_ipc_dispatch_sync` might behave the same as `pcmk_ipc_dispatch_poll` with no timeout.

In theory we could add an asynchronous sleep via a mainloop -- but at that point, we should use `pcmk_ipc_dispatch_main` if a mainloop is available.

Reply:
> `sleep()` would still be synchronous (although it would be less work, at the expense of a delay in finding a message)
> 
> I think `pcmk_poll_ipc()` would be just as busy -- it's just shifting the work to the `poll()` system call. It blocks on `poll()` until either we receive something or the timeout we pass expires.

Blocking on one poll() call is better than constantly iterating it. Since crm_ipc_read() calls qb_ipc_us_ready() -> poll() with a timeout 0,  continuously iterating it without an interval due to EAGAIN significantly raises CPU load.

> I mentioned in T735 that at a glance, it looks like `pcmk_ipc_dispatch_sync` might behave the same as `pcmk_ipc_dispatch_poll` with no timeout.

Indeed.

Reply:
Hmm, the original thinking was that EAGAIN would be a brief condition (often one or two tries). It's definitely a problem that timeouts would have no effect if it's stuck in that loop. It's especially bad if a daemon is waiting on another daemon.

I'm thinking we could poll with a 5s timeout.

An async wait via mainloop might be better, especially for daemons, but it's impractical (every send would need a callback for when it completes), and we'd still need the timeout for non-mainloop sends.

Reply:
Is there anything about this yet by any chance? Otherwise should I try doing something about this? :-)

Reply:
@gao-yan I don't think anything's been started yet

Reply:
Okay, thanks for the update. I'll take a look then.